### PR TITLE
Admin oauth updates

### PR DESCRIPTION
**Pull request by iandouglas**
- admins can see all projects, and note which other user made the project
- admins can delete any student clone for any project
- project clones are sorted by student nickname ascending and clone date descending
- project clones have relative dates/times showing instead of a fixed date
- fixed a bug with stale OAuth tokens, every login will refresh a user's GitHub token

### CODE REVIEW

**Comment by iandouglas** [comment]
already deployed to heroku, but wanted the PR here for visibility

File: app/models/user.rb [hunk 1]
```diff
@@ -3,19 +3,22 @@ class User < ApplicationRecord
...
-      user.name = auth.info.name
-      user.email = auth.info.email
     end
+    user.nickname = auth.info.nickname
+    user.token = auth.credentials.token
+    user.name = auth.info.name
+    user.email = auth.info.email
+    user.save
+    user
   end
```
Comment (iandouglas) [inline app/models/user.rb:23]: moving these pieces out of the find_or_create_by call allows the token/email/etc to be updated every time a user logs into Conveyor Belt

File: app/models/project.rb [hunk 2]
```diff
@@ -25,4 +25,10 @@ def repo_path
 
     [url_segments[1], url_segments[2]].join("/")
   end
+
+  def get_clones
+    clones.joins(:user)
+        .where(project: self)
+        .order('users.nickname asc, users.email asc, clones.created_at desc')
+  end
```
Comment (iandouglas) [inline app/models/project.rb:33]: this allows us to sort the student clones by their name in ascending order, and then by the clone creation in descending order (newest first)

File: app/facades/dashboard_facade.rb [hunk 3]
```diff
@@ -5,6 +5,9 @@ def initialize(user)
 
   def projects
     @projects ||= user.projects
+    if user.admin?
+      @projects = Project.all.order(:project_board_base_url)
+    end
```
Comment (iandouglas) [inline app/facades/dashboard_facade.rb:10]: admins should see all projects, not just the ones they add

File: app/models/user.rb [hunk 4]
```diff
@@ -3,19 +3,22 @@ class User < ApplicationRecord
...
-      user.token = auth.credentials.token
-      user.name = auth.info.name
-      user.email = auth.info.email
     end
+    user.nickname = auth.info.nickname
+    user.token = auth.credentials.token
+    user.name = auth.info.name
+    user.email = auth.info.email
+    user.save
+    user
```
Comment (iandouglas) [inline app/models/user.rb:22]: moving these pieces out of the find_or_create_by allows the email/token/name/etc to be updated every time the user logs in; this fixed the problem for two students yesterday who had stale tokens.

File: app/services/project_board_cloner.rb [hunk 5]
```diff
@@ -84,38 +85,40 @@ def accept_repo_invitation!
...
   def email_student!
+    write_message("sending email to student")
     CloneMailer.with(email: email, clone: clone).send_notification.deliver_now
   end
 
-  def write_message!
-    clone.update(message: message)
+  def write_message(message)
+    @clone.update(message: message)
+    @clone.save!
```
Comment (iandouglas) [inline app/services/project_board_cloner.rb:121]: updated this message writing to update a status for each clone so an admin can see at a glance how things are progressing for a given clone. might be better to number them like "step 5 of 9" or something?

File: app/views/admin/projects/show.html.erb [hunk 6]
```diff
@@ -1,23 +1,23 @@
 <h3><%= @project.name %></h3>
 <ul>
-  <li><strong>Link to send students:</strong> <%= link_to new_project_clone_path(project_id: @project.hash_id), new_project_clone_path(project_id: @project.hash_id) %></li>
+  <li><strong>Link to send students:</strong> <%= link_to new_project_clone_url(project_id: @project.hash_id), new_project_clone_url(project_id: @project.hash_id) %></li>
```
Comment (iandouglas) [inline app/views/admin/projects/show.html.erb:3]: changed this to be the full URL so we can enforce the SSL link on these for students going forward; visiting conveyor belt on a non-SSL URL will block login from working

File: app/views/admin/projects/show.html.erb [hunk 7]
```diff
@@ -1,23 +1,23 @@
...
     <th>Actions</th>
   </tr>
-  <% @project.clones.each do |clone| %>
+  <% @project.get_clones.each do |clone| %>
     <tr>
       <td><%= clone.students %></td>
       <td><%= link_to clone.url, clone.url %></td>
       <td><%= clone.message %></td>
-      <td><%= clone.updated_at %></td>
+      <td><%= distance_of_time_in_words(Time.now, clone.updated_at) %> ago</td>
```
Comment (iandouglas) [inline app/views/admin/projects/show.html.erb:20]: friendlier time distance than a UTC date

File: spec/features/student/student_can_create_a_clone_spec.rb [hunk 8]
```diff
@@ -26,4 +26,35 @@
...
+    click_link "Sign in with GitHub"
+
+    visit admin_project_path(project)
+
+    expect(page).to have_content(clone.url)
+    click_link 'Delete'
+
+    expect(current_path).to eq(admin_project_path(project))
+    expect(page).to_not have_content(clone.url)
+  end
```
Comment (iandouglas) [inline spec/features/student/student_can_create_a_clone_spec.rb:59]: didn't have testing around an admin being able to delete a clone; should also add a test that an admin who didn't create the project can also delete a clone if necessary

File: spec/features/student/student_can_create_a_clone_spec.rb [hunk 9]
```diff
@@ -26,4 +26,35 @@
 
     expect(current_path).to eq(root_path)
   end
+
+  scenario 'admin can delete clone' do
```
Comment (BrianZanti) [inline spec/features/student/student_can_create_a_clone_spec.rb:30]: Could we move this to another file? Maybe a `spec/features/admin/admin_can_delete_a_clone_spec.rb`?

**Pull request merged**